### PR TITLE
Set mapper extra get from address features

### PR DESCRIPTION
**Pull request by mihaicalinluca**

### CODE REVIEW

File: framework/base/src/storage/mappers/queue_mapper.rs [hunk 1]
```diff
@@ -160,13 +172,32 @@ where
         )
     }
 
+    pub fn get_value_from_address(&self, address: &ManagedAddress<SA>, node_id: u32) -> T {
```
Comment (psorinionut) [inline framework/base/src/storage/mappers/queue_mapper.rs:175]: Let's keep the naming convention for public functions.
`get_value_at_address`

File: framework/base/src/storage/mappers/queue_mapper.rs [hunk 2]
```diff
@@ -160,13 +172,32 @@ where
...
+    }
+
     pub fn get_value_option(&self, node_id: u32) -> Option<T> {
         if node_id == NULL_ENTRY {
             return None;
         }
         Some(self.get_value(node_id))
     }
 
+    pub fn get_value_from_address_option(
```
Comment (psorinionut) [inline framework/base/src/storage/mappers/queue_mapper.rs:190]: get_value_at_address_option

File: framework/base/src/storage/mappers/queue_mapper.rs [hunk 3]
```diff
@@ -270,12 +305,20 @@ where
         self.get_value_option(self.get_info().front)
     }
 
+    pub fn front_from_address(&self, address: &ManagedAddress<SA>) -> Option<T> {
```
Comment (psorinionut) [inline framework/base/src/storage/mappers/queue_mapper.rs:308]: `front_at_address`

File: framework/base/src/storage/mappers/queue_mapper.rs [hunk 4]
```diff
@@ -270,12 +305,20 @@ where
...
+        self.get_value_from_address_option(address, self.get_info_from_address(address).front)
+    }
+
     /// Provides a copy to the back element, or `None` if the queue is
     /// empty.
     pub fn back(&self) -> Option<T> {
         self.get_value_option(self.get_info().back)
     }
 
+    pub fn back_from_address(&self, address: &ManagedAddress<SA>) -> Option<T> {
```
Comment (psorinionut) [inline framework/base/src/storage/mappers/queue_mapper.rs:318]: `back_at_address`

File: framework/base/src/storage/mappers/queue_mapper.rs [hunk 5]
```diff
@@ -333,6 +376,10 @@ where
         Iter::new(self)
     }
 
+    pub fn iter_from_address(&self, address: &ManagedAddress<SA>) -> Iter<SA, T> {
```
Comment (psorinionut) [inline framework/base/src/storage/mappers/queue_mapper.rs:379]: `iter_at_address`

File: framework/base/src/storage/mappers/set_mapper.rs [hunk 6]
```diff
@@ -93,6 +101,10 @@ where
         self.queue_mapper.is_empty()
     }
 
+    pub fn is_empty_at_address(&self) -> bool {
+        self.queue_mapper.is_empty()
```
Comment (psorinionut) [inline framework/base/src/storage/mappers/set_mapper.rs:105]: I think here you should use `is_empty_at_address`

File: framework/base/src/storage/mappers/set_mapper.rs [hunk 7]
```diff
@@ -93,6 +101,10 @@ where
...
+    }
+
     /// Returns the number of elements in the set.
     pub fn len(&self) -> usize {
```
Comment (psorinionut) [inline framework/base/src/storage/mappers/set_mapper.rs:109]: Add a `len_at_address` as well.

File: framework/base/src/storage/mappers/set_mapper.rs [hunk 8]
```diff
@@ -114,25 +130,63 @@ where
         self.queue_mapper.get_value_option(next_node_id)
     }
 
+    pub fn next_from_address(&self, address: &ManagedAddress<SA>, value: &T) -> Option<T> {
```
Comment (psorinionut) [inline framework/base/src/storage/mappers/set_mapper.rs:133]: `next_at_address`

File: framework/base/src/storage/mappers/set_mapper.rs [hunk 9]
```diff
@@ -114,25 +130,63 @@ where
...
         }
 
-        let next_node_id = self.queue_mapper.get_node(node_id).previous;
+        let previous_node_id = self.queue_mapper.get_node(node_id).previous;
 
-        self.queue_mapper.get_value_option(next_node_id)
+        self.queue_mapper.get_value_option(previous_node_id)
+    }
+
+    pub fn previous_from_address(&self, address: &ManagedAddress<SA>, value: &T) -> Option<T> {
```
Comment (psorinionut) [inline framework/base/src/storage/mappers/set_mapper.rs:159]: `previous_at_address`

File: framework/base/src/storage/mappers/set_mapper.rs [hunk 10]
```diff
@@ -114,25 +130,63 @@ where
...
+
+        self.queue_mapper
+            .get_value_from_address_option(address, next_node_id)
     }
 
     pub fn front(&self) -> Option<T> {
         self.queue_mapper.front()
     }
 
+    pub fn front_from_address(&self, address: &ManagedAddress<SA>) -> Option<T> {
```
Comment (psorinionut) [inline framework/base/src/storage/mappers/set_mapper.rs:178]: `front_at_address`

File: framework/base/src/storage/mappers/set_mapper.rs [hunk 11]
```diff
@@ -114,25 +130,63 @@ where
...
+        self.queue_mapper.front_from_address(address)
+    }
+
     pub fn back(&self) -> Option<T> {
         self.queue_mapper.back()
     }
 
+    pub fn back_from_address(&self, address: &ManagedAddress<SA>) -> Option<T> {
```
Comment (psorinionut) [inline framework/base/src/storage/mappers/set_mapper.rs:186]: `back_at_address`

File: framework/base/src/storage/mappers/set_mapper.rs [hunk 12]
```diff
@@ -174,15 +228,33 @@ where
...
         self.queue_mapper.iter_from_node_id(node_id)
     }
 
+    pub fn iter_from_value_at_address(
+        &self,
+        address: &ManagedAddress<SA>,
+        value: &T,
+    ) -> Iter<SA, T> {
+        let node_id = self.get_node_id_from_address(address, value);
+        self.queue_mapper.iter_from_node_id(node_id)
```
Comment (psorinionut) [inline framework/base/src/storage/mappers/set_mapper.rs:246]: I think here we need a new addition in the queue_mapper.
`iter_from_node_id_at_address`

**Review by psorinionut — CHANGES_REQUESTED** [review]

**Comment by andrei-marinica** [comment]
Issue resolved in a different way in https://github.com/multiversx/mx-sdk-rs/pull/1384

**Pull request closed without merging**